lib/collection/src/shards/local_shard/clock_map.rs

```diff
@@ -126,15 +128,40 @@ impl ClockMap {
         (is_accepted, new_tick)
     }
 
+    /// Take or clear a snapshot of clocks
+    pub fn snapshot(&mut self, action: ClockMapSnapshot) {
+        match action {
+            ClockMapSnapshot::TakeIfMissing if self.snapshot.is_some() => {}
+            ClockMapSnapshot::Clear if self.snapshot.is_none() => {}
+
+            ClockMapSnapshot::TakeIfMissing => {
+                self.snapshot.replace(self.clocks.clone());
+                self.changed = true;
+            }
+            ClockMapSnapshot::Clear => {
+                self.snapshot.take();
+                self.changed = true;
+            }
+        }
+    }
+
     /// Create a recovery point based on the current clock map state, so that we can recover any
     /// new operations with new clock values
     ///
+    /// If `use_snapshot` is `true`, the recovery point is created from the clocks snapshot if it
+    /// exists. Otherwise the current clocks are used.
+    ///
     /// The recovery point contains every clock that is in this clock map. So, it represents all
     /// the clock ticks we have.
-    pub fn to_recovery_point(&self) -> RecoveryPoint {
+    pub fn to_recovery_point(&self, use_snapshot: bool) -> RecoveryPoint {
+        let clocks = if use_snapshot {
+            self.snapshot.as_ref().unwrap_or(&self.clocks)
+        } else {
+            &self.clocks
+        };
```

Comment:
I'd remove `use_snapshot` parameter and just always do `self.snapshot.as_ref().unwrap_or(&self.clocks)`.

Reply:
Implemented in <https://github.com/qdrant/qdrant/pull/7791/commits/854fe281be94ac54d773cd4d5ab462a80372801d>